airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py

```diff
@@ -0,0 +1,117 @@
+            self.config["data_center"] = data_center
+
+    @staticmethod
+    def get_oauth_data_center(access_token: str) -> str:
+        """
+        Every Mailchimp API request must be sent to a specific data center.
+        The data center is already embedded in API keys, but not OAuth access tokens.
+        This method retrieves the data center for OAuth credentials.
+        """
+        try:
+            response = requests.get(
+                "https://login.mailchimp.com/oauth2/metadata", headers={"Authorization": "OAuth {}".format(access_token)}
+            )
+
+            # Requests to this endpoint will return a 200 status code even if the access token is invalid.
+            error = response.json().get("error")
+            if error == "invalid_token":
+                raise ValueError("The access token you provided was invalid. Please check your credentials and try again.")
+            return response.json()["dc"]
+
+        # Handle any other exceptions that may occur.
+        except Exception as e:
+            raise Exception(f"An error occured while retrieving the data center for your account. \n {repr(e)}")
+
+
+class MailChimpAuthenticator(DeclarativeAuthenticator):
+    config: Mapping[str, Any]
+    bearer: BearerAuthenticator
+    basic: BasicHttpAuthenticator
+
+    def __new__(cls, bearer, basic, config, *args, **kwargs):
+        if config.get("credentials", {}).get("auth_type") == "oauth2.0":
+            return bearer
+        else:
+            return basic
+
+
+class MailChimpRecordFilter(RecordFilter):
+    """
+    Filter applied on a list of Records.
+    """
+
+    def filter_records(
+        self,
+        records: List[Mapping[str, Any]],
+        stream_state: StreamState,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> List[Mapping[str, Any]]:
+        current_state = [x for x in stream_state.get("states", []) if x["partition"]["id"] == stream_slice.partition["id"]]
+        # TODO: REF what to do if no start_date mentioned (see manifest)
+        #  implement the same logic
+        start_date = self.config.get("start_date", (pendulum.now() - pendulum.duration(days=700)).to_iso8601_string())
+        if current_state and start_date:
+            filter_value = max(start_date, current_state[0]["cursor"][self.parameters["cursor_field"]])
+            return [record for record in records if record[self.parameters["cursor_field"]] > filter_value]
+        return records
+
+
+class MailChimpRecordFilterEmailActivity(RecordFilter):
+    def filter_records(
+        self,
+        records: List[Mapping[str, Any]],
+        stream_state: StreamState,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> List[Mapping[str, Any]]:
+
+        return [{**record, **activity_item} for record in records for activity_item in record.pop("activity", [])]
```

Comment:
Question so I learn: looks like this filter takes a dict of record and returns a separate row for each record and each activity in that record. 

If there are no activities at all, the record will be filtered out. If there are multiple activities on a single record (if that is possible), that will return multiple row for each found activity item. 

How wrong am I?

If that's correct, than this thing both _filters_ records, but also _transforms_ them and generates them, right?

Reply:
why is this defined as being a filter? Like Natik mentioned, I would expect this to be a custom transformation followed by a filter that checks if there were activities.

Separating the two concerns allows us to extract some logic that can be generalized.

Instead of creating a custom filter, we should introduce a flatten_field transformation which takes a nested object and brings it to the root

The transformation:
```
{
"key": 123,
"activities": {"nested": "value", "another_nested": "value2"} 
}

to 
{
"key": 123,
"nested:" value",
"another_nested": "value2"
}
```

would then be described as
```
transformations:
  type: FlattenField
  field_pointers:
    - activities
```

Here is a tentative schema definition:
```
  FlattenFields:
    title: Flatten Fields
    type: object
    required:
      - type
      - field_pointers
    properties:
      type:
        type: string
        enum: [FlattenFields]
      field_pointers:
        title: Field Paths
        type: array
        items:
          items:
            type: string
        examples:
          - ["activities"]
```

Reply:
Hmm, I can agree that this is more about `transformation` rather than `filtering`, but 

```json
{
"key": 123,
"activities": [
{"nested": "value"},
 {"another_nested": "value2"}]
}

```
should be transformed to 2 records as `activities` is a list :
```json
[
  {
    "key": 123,
    "nested": "value"
  },
  {
    "key": 123,
    "another_nested": "value2"
  }
]
```
it means that we will have more records than original list, while current implementation does not expect the number of records to change:

https://github.com/airbytehq/airbyte/blob/4061f08f3d34e2de651c8d2a445952f3335ebcc9/airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_selector.py#L88-L97

Reply:
got it! I did misunderstand. 

Can you make a proposal for how you'd want to use this feature? When considering new features or custom components, it's useful to have a description, examples, and a desired interface (when known) to help us understand the use case. 

Then we can implement and make any required changes to the record transform interface

Reply:
Sure, I'll get back with all propositions as soon as this PR will be in review, I'm still working on it.